nixos/modules/security/agnos.nix

```
@@ -0,0 +1,306 @@
+              (subject to provider's rate limits,
+              see e.g. [Let's Encrypt Rate Limits](https://letsencrypt.org/docs/rate-limits/)).
+            '';
+          };
+        };
+      };
+    };
+
+    generateKeys = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable automatic generation of account keys.
+
+          When this is `true`, a key will be generated for each account where
+          the file referred to by the `private_key` path does not exist yet.
+
+          Currently, only RSA keys can be generated.
+        '';
+      };
+
+      keySize = mkOption {
+        type = types.int;
+        default = 4096;
+        description = ''
+          Key size in bits to use when generating new keys.
+        '';
+      };
+    };
+
+    server = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = ''
+        ACME Directory Resource URI. Defaults to Let's Encrypt's production endpoint,
+        `https://acme-v02.api.letsencrypt.org/directory`, if unset.
+      '';
+    };
+
+    serverCa = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        The root certificate (in PEM format) of the ACME server's HTTPS interface.
+      '';
+    };
+
+    persistent = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        When `true`, use a persistent systemd timer.
+      '';
+    };
+
+    startAt = mkOption {
+      type = types.either types.str (types.listOf types.str);
+      default = "daily";
+      example = "02:00";
+      description = ''
+        How often or when to run agnos.
+
+        The format is described in
+        {manpage}`systemd.time(7)`.
+      '';
+    };
+
+    temporarilyOpenFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        When `true`, will open the port specified in `settings.dns_listen_addr`
+        before running the agnos service, and close it when agnos finishes running.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = name;
+      description = ''
+        Group to run Agnos as. The acquired certificates will be owned by this group.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = name;
+      description = ''
+        User to run Agnos as. The acquired certificates will be owned by this user.
+      '';
+    };
+  };
+
+  config =
+    let
+      configFile = format.generate "agnos.toml" cfg.settings;
+      port = lib.toInt (lib.last (builtins.split ":" cfg.settings.dns_listen_addr));
+
+      useNftables = config.networking.nftables.enable;
+
+      # nftables implementation for temporarilyOpenFirewall
+      nftablesSetup = pkgs.writeShellScript "agnos-fw-setup" ''
+        ${pkgs.nftables}/bin/nft add element inet nixos-fw temp-ports "{ tcp . ${toString port} }"
```

```suggestion
        ${lib.getExe pkgs.nftables} add element inet nixos-fw temp-ports "{ tcp . ${toString port} }"
```
And similar below. `${lib.getExe' iptables "iptables"}` in helpers.